src/ILLink.RoslynAnalyzer/RequiresUnreferencedCodeAnalyzer.cs

```diff
@@ -31,6 +33,49 @@ public sealed class RequiresUnreferencedCodeAnalyzer : RequiresAnalyzerBase
 				operationContext.Operation.Syntax.GetLocation ()));
 		};
 
+		[SuppressMessage ("MicrosoftCodeAnalysisPerformance", "RS1008",
+			Justification = "This action is registered through a compilation start action, so that the instances which " +
+			"can register this operation action will not outlive a compilation's lifetime, avoiding the possibility of " +
+			"this data causing stale compilations to remain in memory.")]
+		static readonly Action<OperationAnalysisContext> s_constructorConstraint = operationContext => {
+			if (FindContainingSymbol (operationContext, DiagnosticTargets.All) is not ISymbol containingSymbol ||
+			containingSymbol.HasAttribute (RequiresUnreferencedCodeAttribute))
+				return;
+
+			var typeParams = ImmutableArray<ITypeParameterSymbol>.Empty;
+			var typeArgs = ImmutableArray<ITypeSymbol>.Empty;
+			if (operationContext.Operation is IObjectCreationOperation objectCreationOperation &&
+				objectCreationOperation.Type is INamedTypeSymbol objectType && objectType.IsGenericType) {
+				typeParams = objectType.TypeParameters;
+				typeArgs = objectType.TypeArguments;
```

Comment:
This only reacts to calling a constructor on a generic type.
Code like this should still warn though:
```C#
static class MyTest<T> where T : new ()
{
    static void DoNothing() {}
}

void Test()
{
    // IL2026
    MyTest.DoNothing<ClassWithRUCOnCtor>();
}
```

I must admit that I don't know enough about analyzers to tell what type of solution would be the best here. Linker will warn anywhere the generic is instantiated. Even cases like `typeof(MyType<ClassWithRUCOnCtor>)` will generate the warning.

This is obviously not that common, so it's OK to solve this later...

Reply:
I moved the analysis context used here to a `SyntaxNodeAnalysisContext`; this way we look for syntax nodes that represent generic names and check, for their type parameters, if these have any constructor constraints. I think this would cover up all cases. I've added the scenario you outline above (and the analyzer warns now).